spec/index.bs

```diff
@@ -6954,8 +7027,8 @@ interface GPUQueue {
       GPUImageDataLayout dataLayout,
       GPUExtent3D size);
 
-    undefined copyImageBitmapToTexture(
-        GPUImageCopyImageBitmap source,
+    undefined copyExternalImageToTexture(
+        GPUImageCopyExternalImage source,
```

Comment:
(Inspired by the sample [here](https://source.chromium.org/chromium/chromium/src/+/master:content/test/data/gpu/vc/webgpu_videos_mxn.html) but don't think we need an issue to discuss, maybe)
Is it possible that we accept the array of GPUImageCopyExternalImage and with the array of CopySize correspond to each other. This will make this API have a stiching semantic(from video view) and avoid some impl overhead(e.g. extra sync token)

To describe it more detail:
In previous WebGL/WebGPU, if developer upload many videos to WebGPU and rendering them all into one canvas,
They need to do many CopyExternalImageToTexture/TexImage2D, get many textures, binding all of these textures and start a rendering pipeline to draw them on the canvas.

But if CopyExternalImageToTexture accept arrays of GPUImageCopyExternalImage, the developer will handle this case to :
Call copyExternalImageToTexture once with arrays of GPUImageCopyExternalImage, bind that texture and render it into canvas.

But this will add the complex of the API and validation.

Not sure whether this could be an wroth opt. If it is, we could file an issue to record it.

Reply:
I don't think this case is so common it's worth optimizing the API surface for. I think this really just saves some bindings call cost. Shouldn't it be possible to optimize away the extra sync tokens by recognizing when multiple of these copy commands are issued in a row?

> get many textures, binding all of these textures and start a rendering pipeline to draw them on the canvas.

If the videos are of consistent sizes, then this isn't necessary: the user can just copy multiple videos into a single texture at different offsets.

However if the videos are of inconsistent sizes, then an array function won't help, we would need to support resizing (blitting). To be honest, maybe we need to add resizing behavior here: for videos, it's common that a video source's resolution varies over time, or multiple video sources have different resolutions but need to be displayed at the same size (e.g. in a video call, the high-res outgoing video + 20 low-res incoming videos).

Reply:
To be clear, we probably don't need resizing behavior here immediately. If an app just wanted to directly paste a bunch of videos into a grid, they could do it with 2d canvas. If it wants to do anything more complicated, it needs the videos as separate textures anyway, and those textures just be imported GPUExternalTextures (or can be regular GPUTextures with the same resolutions as the videos).

Reply:
> If the videos are of consistent sizes, then this isn't necessary: the user can just copy multiple videos into a single texture at different offsets.

Yes, I think this is the common behaviour. My case is not reasonable.

> Shouldn't it be possible to optimize away the extra sync tokens by recognizing when multiple of these copy commands are issued in a row

Might be hard to do this. This means that all copy ops are delay executing. But sync token issue is specific for browser and maybe some experts on this knows the better solution.

> To be clear, we probably don't need resizing behavior here immediately. If an app just wanted to directly paste a bunch of videos into a grid, they could do it with 2d canvas. If it wants to do anything more complicated, it needs the videos as separate textures anyway, and those textures just be imported GPUExternalTextures (or can be regular GPUTextures with the same resolutions as the videos).

I tend to left this handling by developers too. Developers could monitor the video content size with attribute [`HTMLVideoElement.videoHeight`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLVideoElement/videoHeight) and [`HTMLVideoElement.videoWidth`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLVideoElement/videoWidth). With the copy API only, they could prepare different size texture and then apply their own resize ops. This would saving define _default_ _resize_ behaviour in spec.